File: samples/java/Bingo/Client/src/main/java/systems/zlink/samples/bingo/client/BingoClientScenario.java
package systems.zlink.samples.bingo.client;

import systems.zlink.samples.bingo.client.configuration.SampleNames;
import systems.zlink.samples.bingo.shared.contracts.BingoMessages;
import systems.zlink.samples.bingo.shared.contracts.Messages;
import systems.zlink.stream.connector.ZLinkStreamConnector;
import systems.zlink.stream.connector.ZLinkStreamMessage;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.atomic.AtomicInteger;

public final class BingoClientScenario {
    public void run(
            ZLinkStreamConnector client1,
            ZLinkStreamConnector client2,
            ZLinkStreamConnector observer)
            throws Exception {
        client1.connect().submit().toCompletableFuture().join();
        client2.connect().submit().toCompletableFuture().join();
        observer.connect().submit().toCompletableFuture().join();

        Messages.AuthenticateRes client1Auth =
                client1.request(BingoMessages.authenticateReq("player-1"))
                        .submit(Messages.AuthenticateRes.class)
                        .toCompletableFuture()
                        .join();
        ensure(client1Auth.getActorId().equals("player-1"));
        AtomicInteger client1OwnJoinNotifications = new AtomicInteger();
        client1.on(
                SampleNames.PlayerJoinedPacket,
                Messages.PlayerJoinedNotify.class,
                message -> {
                    if (client1Auth.getActorId().equals(message.payload().getActorId())) {
                        client1OwnJoinNotifications.incrementAndGet();
                    }
                    return CompletableFuture.completedFuture(null);
                });

        Messages.MatchBingoRes client1Match =
                client1.request(BingoMessages.matchBingoReq("two-player"))
                        .submit(Messages.MatchBingoRes.class)
                        .toCompletableFuture()
                        .join();
        ensure(client1Match.getState().getStatus().equals("WaitingForPlayers"));
        ensure(client1Match.getState().getHostActorId().equals(client1Auth.getActorId()));

        Messages.AuthenticateRes observerAuth =
                observer.request(BingoMessages.authenticateReq("observer"))
                        .submit(Messages.AuthenticateRes.class)
                        .toCompletableFuture()
                        .join();
        ensure(observerAuth.getActorId().equals("observer"));
        Messages.ObserveBingoEventsRes observed =
                observer.request(BingoMessages.observeBingoEventsReq(client1Match.getRoomId()))
                        .submit(Messages.ObserveBingoEventsRes.class)
                        .toCompletableFuture()
                        .join();
        ensure(observed.getSubscribed());
        // Register the observer wait before the second player can start the draw loop.
        var rewardAnnounced =
                observer.waitFor(SampleNames.RewardAnnouncedPacket)
                        .where(
                                Messages.BingoRewardAnnouncedNotify.class,
                                message ->
                                        message.payload()
                                                .getRoomId()
                                                .equals(client1Match.getRoomId()))
                        .submit(Messages.BingoRewardAnnouncedNotify.class);

        var client1SawClient2Join =
                client1.waitFor(SampleNames.PlayerJoinedPacket)
                        .submit(Messages.PlayerJoinedNotify.class);
        var client1Started =
                client1.waitFor(SampleNames.GameStartedPacket)
                        .submit(Messages.BingoGameStartedNotify.class);
        var client2Started =
                client2.waitFor(SampleNames.GameStartedPacket)
                        .submit(Messages.BingoGameStartedNotify.class);

        Messages.AuthenticateRes client2Auth =
                client2.request(BingoMessages.authenticateReq("player-2"))
                        .submit(Messages.AuthenticateRes.class)
                        .toCompletableFuture()
                        .join();
        ensure(client2Auth.getActorId().equals("player-2"));
        ensure(!client2Auth.getActorId().equals(client1Auth.getActorId()));
        AtomicInteger client2OwnJoinNotifications = new AtomicInteger();
        client2.on(
                SampleNames.PlayerJoinedPacket,
                Messages.PlayerJoinedNotify.class,
                message -> {
                    if (client2Auth.getActorId().equals(message.payload().getActorId())) {
                        client2OwnJoinNotifications.incrementAndGet();
                    }
                    return CompletableFuture.completedFuture(null);
                });

        Messages.MatchBingoRes client2Match =
                client2.request(BingoMessages.matchBingoReq("two-player"))
                        .submit(Messages.MatchBingoRes.class)
                        .toCompletableFuture()
                        .join();
        ensure(client2Match.getRoomId().equals(client1Match.getRoomId()));
        ensure(client2Match.getState().getStatus().equals("WaitingForPlayers"));

        Messages.PlayerJoinedNotify join =
                client1SawClient2Join.toCompletableFuture().join().payload();
        ensure(join.getActorId().equals(client2Auth.getActorId()));
        System.out.println("stream-handler sample=Bingo client=player1 message=PlayerJoinedNotify");
        ensure(
                client1Started
                        .toCompletableFuture()
                        .join()
                        .payload()
                        .getState()
                        .getStatus()
                        .equals("Running"));
        ensure(
                client2Started
                        .toCompletableFuture()
                        .join()
                        .payload()
                        .getState()
                        .getStatus()
                        .equals("Running"));

        Messages.SubmitBingoCardRes client2Card =
                client2.request(
                                BingoMessages.submitBingoCardReq(
                                        client2Match.getRoomId(), BingoClientCards.Player2))
                        .submit(Messages.SubmitBingoCardRes.class)
                        .toCompletableFuture()
                        .join();
        ensure(client2Card.getState().getStatus().equals("Running"));
        ensure(
                client2Card.getState().getPlayersList().stream()
                        .anyMatch(
                                player ->
                                        player.getActorId().equals(client2Auth.getActorId())
                                                && player.getCardCount() == 9));
        var client1Ended =
                client1.waitFor(SampleNames.GameEndedPacket)
                        .submit(Messages.BingoGameEndedNotify.class);
        var client2Ended =
                client2.waitFor(SampleNames.GameEndedPacket)
                        .submit(Messages.BingoGameEndedNotify.class);
        List<CompletionStage<ZLinkStreamMessage<Messages.BingoNumberDrawnNotify>>> client1Draws =
                new ArrayList<>();
        List<CompletionStage<ZLinkStreamMessage<Messages.BingoNumberDrawnNotify>>> client2Draws =
                new ArrayList<>();
        for (int drawSeq = 1; drawSeq <= 15; drawSeq++) {
            int expectedDrawSeq = drawSeq;
            client1Draws.add(
                    client1.waitFor(SampleNames.NumberDrawnPacket)
                            .where(
                                    Messages.BingoNumberDrawnNotify.class,
                                    message -> message.payload().getDrawSeq() == expectedDrawSeq)
                            .submit(Messages.BingoNumberDrawnNotify.class));
            client2Draws.add(
                    client2.waitFor(SampleNames.NumberDrawnPacket)
                            .where(
                                    Messages.BingoNumberDrawnNotify.class,
                                    message -> message.payload().getDrawSeq() == expectedDrawSeq)
                            .submit(Messages.BingoNumberDrawnNotify.class));
        }

        Messages.SubmitBingoCardRes client1Card =
                client1.request(
                                BingoMessages.submitBingoCardReq(
                                        client1Match.getRoomId(), BingoClientCards.Player1))
                        .submit(Messages.SubmitBingoCardRes.class)
                        .toCompletableFuture()
                        .join();
        ensure(client1Card.getState().getStatus().equals("Running"));
        ensure(client1Card.getState().getPlayersCount() == 2);
        ensure(
                client1Card.getState().getPlayersList().stream()
                        .allMatch(player -> player.getCardCount() == 9));

        List<Messages.BingoNumberDrawnNotify> drawnNumbers = new ArrayList<>();
        for (int drawSeq = 1; drawSeq <= 15; drawSeq++) {
            Messages.BingoNumberDrawnNotify client1Drawn =
                    client1Draws.get(drawSeq - 1).toCompletableFuture().join().payload();
            Messages.BingoNumberDrawnNotify client2Drawn =
                    client2Draws.get(drawSeq - 1).toCompletableFuture().join().payload();
            drawnNumbers.add(client1Drawn);
            ensure(client1Drawn.getDrawSeq() == drawSeq);
            ensure(client2Drawn.getDrawSeq() == drawSeq);
            ensure(client2Drawn.getNumber() == client1Drawn.getNumber());
            ensure(client2Drawn.getState().equals(client1Drawn.getState()));

            if (client1Drawn.getState().getStatus().equals("Finished")) {
                break;
            }
        }
        ensure(!drawnNumbers.isEmpty());
        ensure(drawnNumbers.getLast().getState().getStatus().equals("Finished"));

        Messages.BingoRoomState client1Result =
                client1Ended.toCompletableFuture().join().payload().getState();
        Messages.BingoRoomState client2Result =
                client2Ended.toCompletableFuture().join().payload().getState();
        ensure(client1Result.getStatus().equals("Finished"));
        ensure(client2Result.getStatus().equals("Finished"));
        ensure(client2Result.getDrawnNumbersList().equals(client1Result.getDrawnNumbersList()));
        ensure(client2Result.getWinnersList().equals(client1Result.getWinnersList()));
        ensure(
                client2Result.getPlayersList().stream()
                        .map(Messages.BingoPlayerState::getActorId)
                        .toList()
                        .equals(
                                client1Result.getPlayersList().stream()
                                        .map(Messages.BingoPlayerState::getActorId)
                                        .toList()));
        ensure(
                client1Result
                        .getDrawnNumbersList()
                        .equals(
                                drawnNumbers.stream()
                                        .map(Messages.BingoNumberDrawnNotify::getNumber)
                                        .toList()));
        ensure(client1Result.getWinnersList().equals(List.of(client1Auth.getActorId())));
        ensure(
                client1Result.getPlayersList().stream()
                        .allMatch(player -> player.getCardCount() == 9));
        ensure(client1Result.getPlayersList().stream().allMatch(player -> player.getMarks(4)));

        Messages.BingoRewardAnnouncedNotify reward =
                rewardAnnounced.toCompletableFuture().join().payload();
        ensure(reward.getActorId().equals(client1Auth.getActorId()));
        ensure(reward.getItemId().equals("rare-golden-dauber"));
        ensure(reward.getItemName().equals("Golden Dauber"));
        ensure(reward.getRarity().equals("Legendary"));

        Messages.StopObservingBingoEventsRes stopped =
                observer.request(
                                BingoMessages.stopObservingBingoEventsReq(client1Match.getRoomId()))
                        .submit(Messages.StopObservingBingoEventsRes.class)
                        .toCompletableFuture()
                        .join();
        ensure(stopped.getStopped());
        ensure(client1OwnJoinNotifications.get() == 0);
        ensure(client2OwnJoinNotifications.get() == 0);
    }

    private static void ensure(boolean condition) {
        if (!condition) {
            throw new IllegalStateException("Ensure failed");
        }
    }

    private static final class BingoClientCards {
        private static final List<Integer> Player1 = List.of(1, 2, 3, 4, 0, 6, 7, 8, 9);
        private static final List<Integer> Player2 = List.of(10, 11, 12, 13, 0, 14, 4, 5, 6);
    }
}

File: samples/kotlin/Bingo/Client/src/main/kotlin/systems/zlink/samples/kotlin/bingo/client/BingoClientScenario.kt
package systems.zlink.samples.kotlin.bingo.client

import java.time.Duration
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import systems.zlink.framework.kotlin.ZLinkKotlinStreamConnector
import systems.zlink.framework.kotlin.await
import systems.zlink.framework.kotlin.awaitReply
import systems.zlink.samples.kotlin.bingo.client.configuration.SampleNames
import systems.zlink.samples.kotlin.bingo.shared.contracts.AuthenticateReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.AuthenticateRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoGameEndedNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoGameStartedNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoNumberDrawnNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRewardAnnouncedNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.ObserveBingoEventsReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.ObserveBingoEventsRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.PlayerJoinedNotify
import systems.zlink.samples.kotlin.bingo.shared.contracts.StopObservingBingoEventsReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.StopObservingBingoEventsRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.SubmitBingoCardReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.SubmitBingoCardRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.card
import systems.zlink.samples.kotlin.bingo.shared.contracts.drawnNumbers
import systems.zlink.samples.kotlin.bingo.shared.contracts.marks
import systems.zlink.samples.kotlin.bingo.shared.contracts.players
import systems.zlink.samples.kotlin.bingo.shared.contracts.winners

class BingoClientScenario {
    suspend fun run(
        client1: ZLinkKotlinStreamConnector,
        client2: ZLinkKotlinStreamConnector,
        observer: ZLinkKotlinStreamConnector,
    ) = coroutineScope {
        client1.connect().await()
        client2.connect().await()
        observer.connect().await()
        val client1Auth = client1.request(AuthenticateReq("player-1")).awaitReply<AuthenticateRes>()
        ensure(client1Auth.actorId == "player-1")

        val client1NoSelfJoin =
            async(start = CoroutineStart.UNDISPATCHED) {
                client1
                    .expectNone<PlayerJoinedNotify>(SampleNames.PlayerJoinedPacket)
                    .within(Duration.ofMillis(400))
                    .await()
            }
        val client1Match = client1.request(MatchBingoReq("two-player")).awaitReply<MatchBingoRes>()
        ensure(client1Match.state.status == "WaitingForPlayers")
        ensure(client1Match.state.hostActorId == client1Auth.actorId)
        client1NoSelfJoin.await()

        val observerAuth =
            observer.request(AuthenticateReq("observer")).awaitReply<AuthenticateRes>()
        ensure(observerAuth.actorId == "observer")
        val observed =
            observer
                .request(ObserveBingoEventsReq(client1Match.roomId))
                .awaitReply<ObserveBingoEventsRes>()
        ensure(observed.subscribed)
        // Register the observer wait before the second player can start the draw loop.
        val rewardAnnounced =
            observer
                .waitFor<BingoRewardAnnouncedNotify>()
                .where { message -> message.payload().roomId == client1Match.roomId }
                .let { wait -> async(start = CoroutineStart.UNDISPATCHED) { wait.await() } }

        val client1SawClient2Join =
            client1
                .waitFor<PlayerJoinedNotify>()
                .where { message -> message.payload().actorId == "player-2" }
                .let { wait -> async(start = CoroutineStart.UNDISPATCHED) { wait.await() } }
        val client1Started =
            async(start = CoroutineStart.UNDISPATCHED) {
                client1.waitFor<BingoGameStartedNotify>().await()
            }
        val client2Started =
            async(start = CoroutineStart.UNDISPATCHED) {
                client2.waitFor<BingoGameStartedNotify>().await()
            }

        val client2Auth = client2.request(AuthenticateReq("player-2")).awaitReply<AuthenticateRes>()
        ensure(client2Auth.actorId == "player-2")
        ensure(client2Auth.actorId != client1Auth.actorId)

        val client2NoSelfJoin =
            async(start = CoroutineStart.UNDISPATCHED) {
                client2
                    .expectNone<PlayerJoinedNotify>(SampleNames.PlayerJoinedPacket)
                    .within(Duration.ofMillis(400))
                    .await()
            }
        val client2Match = client2.request(MatchBingoReq("two-player")).awaitReply<MatchBingoRes>()
        ensure(client2Match.roomId == client1Match.roomId)
        ensure(client2Match.state.status == "WaitingForPlayers")

        val join = client1SawClient2Join.await().payload()
        ensure(join.actorId == client2Auth.actorId)
        ensure(join.roomId == client1Match.roomId)
        ensure(
            join.state.players.map { player -> player.actorId }.toSet() ==
                setOf(client1Auth.actorId, client2Auth.actorId)
        )
        println("stream-handler sample=Bingo client=player1 message=PlayerJoinedNotify")
        client2NoSelfJoin.await()
        ensure(client1Started.await().payload().state.status == "Running")
        ensure(client2Started.await().payload().state.status == "Running")

        val client2Card =
            client2
                .request(SubmitBingoCardReq(client2Match.roomId, BingoClientCards.Player2))
                .awaitReply<SubmitBingoCardRes>()
        ensure(client2Card.state.status == "Running")
        ensure(
            client2Card.state.players
                .single { player -> player.actorId == client2Auth.actorId }
                .card
                .size == 9
        )

        val client1Ended =
            async(start = CoroutineStart.UNDISPATCHED) {
                client1.waitFor<BingoGameEndedNotify>().await()
            }
        val client2Ended =
            async(start = CoroutineStart.UNDISPATCHED) {
                client2.waitFor<BingoGameEndedNotify>().await()
            }

        val client1Draws =
            (1..15).map { expectedDrawSeq ->
                async(start = CoroutineStart.UNDISPATCHED) {
                    client1
                        .waitFor<BingoNumberDrawnNotify>()
                        .where { message -> message.payload().drawSeq == expectedDrawSeq }
                        .await()
                }
            }
        val client2Draws =
            (1..15).map { expectedDrawSeq ->
                async(start = CoroutineStart.UNDISPATCHED) {
                    client2
                        .waitFor<BingoNumberDrawnNotify>()
                        .where { message -> message.payload().drawSeq == expectedDrawSeq }
                        .await()
                }
            }

        val client1Card =
            client1
                .request(SubmitBingoCardReq(client1Match.roomId, BingoClientCards.Player1))
                .awaitReply<SubmitBingoCardRes>()
        ensure(client1Card.state.status == "Running")
        ensure(client1Card.state.players.size == 2)
        ensure(client1Card.state.players.all { player -> player.card.size == 9 })

        val drawnNumbers = mutableListOf<BingoNumberDrawnNotify>()
        for (drawSeq in 1..15) {
            val client1Drawn = client1Draws[drawSeq - 1].await().payload()
            val client2Drawn = client2Draws[drawSeq - 1].await().payload()
            drawnNumbers += client1Drawn
            ensure(client1Drawn.drawSeq == drawSeq)
            ensure(client2Drawn.drawSeq == drawSeq)
            ensure(client2Drawn.number == client1Drawn.number)
            ensure(client2Drawn.state == client1Drawn.state)

            if (client1Drawn.state.status == "Finished") {
                break
            }
        }
        ensure(drawnNumbers.isNotEmpty())
        ensure(drawnNumbers.last().state.status == "Finished")
        client1Draws.drop(drawnNumbers.size).forEach { wait -> wait.cancel() }
        client2Draws.drop(drawnNumbers.size).forEach { wait -> wait.cancel() }

        val client1Result = client1Ended.await().payload().state
        val client2Result = client2Ended.await().payload().state
        ensure(client1Result.status == "Finished")
        ensure(client2Result.status == "Finished")
        ensure(client2Result.drawnNumbers == client1Result.drawnNumbers)
        ensure(client2Result.winners == client1Result.winners)
        ensure(
            client2Result.players.map { player -> player.actorId } ==
                client1Result.players.map { player -> player.actorId }
        )
        ensure(client1Result.drawnNumbers == drawnNumbers.map { notify -> notify.number })
        ensure(client1Result.winners == listOf(client1Auth.actorId))
        ensure(client1Result.players.all { player -> player.card.size == 9 })
        ensure(client1Result.players.all { player -> player.marks[4] })

        val reward = rewardAnnounced.await().payload()
        ensure(reward.roomId == client1Match.roomId)
        ensure(reward.actorId == client1Auth.actorId)
        ensure(reward.drawSeq == drawnNumbers.last().drawSeq)
        ensure(reward.itemId == "rare-golden-dauber")
        ensure(reward.itemName == "Golden Dauber")
        ensure(reward.rarity == "Legendary")

        val stopped =
            observer
                .request(StopObservingBingoEventsReq(client1Match.roomId))
                .awaitReply<StopObservingBingoEventsRes>()
        ensure(stopped.stopped)
        println("bingo=completed")
    }
}

private fun ensure(condition: Boolean) {
    if (!condition) {
        throw IllegalStateException("Ensure failed")
    }
}

private object BingoClientCards {
    val Player1: List<Int> = listOf(1, 2, 3, 4, 0, 6, 7, 8, 9)
    val Player2: List<Int> = listOf(10, 11, 12, 13, 0, 14, 4, 5, 6)
}
